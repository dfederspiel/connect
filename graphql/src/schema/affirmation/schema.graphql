type Affirmation {
  from: ID!
  to: ID!
}

extend type Mutation {
  sendAffirmation(userId: ID!): Affirmation
}

extend type Subscription {
  affirmationGiven: Affirmation
}
