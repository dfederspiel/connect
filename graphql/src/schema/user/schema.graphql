type User {
  id: ID!
  email: String
  domain: String
}

extend type Query {
  user: User
  users: [User]
}
