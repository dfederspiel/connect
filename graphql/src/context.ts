import { PrismaClient } from '@prisma/client';
import { mockDeep } from 'jest-mock-extended';
import { DeepMockProxy } from 'jest-mock-extended/lib/mjs/Mock';

export type Context = {
  prisma: PrismaClient;
};

export type MockContext = {
  prisma: DeepMockProxy<PrismaClient>;
};

export const createMockContext = (): MockContext => {
  return {
    prisma: mockDeep<PrismaClient>(),
  };
};
