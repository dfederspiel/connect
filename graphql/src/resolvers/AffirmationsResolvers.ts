import { User } from '@prisma/client';
import { gql } from 'apollo-server-express';
import { withFilter } from 'graphql-subscriptions';
import { IResolvers } from '@graphql-tools/utils';
import { IPubSub } from '../lib/types';

const AFFIRMATION_GIVEN = 'AFFIRMATION_GIVEN';

export const AffirmationsTypeDefs = gql`
  type Affirmation {
    from: ID!
    to: ID!
  }

  extend type Mutation {
    sendAffirmation(userId: ID!): Affirmation
  }

  extend type Subscription {
    affirmationGiven: Affirmation
  }
`;

interface IAffirmationsMutationResolvers {
  sendAffirmation(
    parent: any,
    args: any,
    context: any,
    other: any,
  ): Promise<{
    from: number;
    to: any;
  }>;
}

interface IAffirmationsSubscriptionResolvers {
  affirmationGiven: {
    subscribe(): any;
  };
}
interface IAffirmationsResolvers extends IResolvers {
  Mutation: IAffirmationsMutationResolvers;
  Subscription: IAffirmationsSubscriptionResolvers;
}

export default class AffirmationsResolvers {
  resolvers: IAffirmationsResolvers;
  pubsub: IPubSub;

  constructor(pubsub: IPubSub) {
    this.resolvers = this.initializeResolvers();
    this.pubsub = pubsub;
  }

  initializeResolvers = (): IAffirmationsResolvers => {
    return {
      Mutation: {
        sendAffirmation: async (_: any, { userId }: any, context: { user: User }) => {
          console.log('WEBSOCKET SEND', _, userId, context.user);

          const affirmation = {
            from: context.user.id,
            to: userId,
          };
          this.pubsub.publish(AFFIRMATION_GIVEN, affirmation);
          return affirmation;
        },
      },
      Subscription: {
        affirmationGiven: {
          subscribe: withFilter(
            () => this.pubsub.asyncIterator('AFFIRMATION_GIVEN'),
            (payload, _args, context) => {
              console.log('WEBSOCKET SUBSCRIBE', payload, _args, context);
              return parseInt(payload.to) === context.user.id;
            },
          ),
        },
      },
    };
  };
}
