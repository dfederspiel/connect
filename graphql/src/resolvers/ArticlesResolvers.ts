import { IResolvers } from '@graphql-tools/utils';
import { Article, User } from '@prisma/client';
import { gql } from 'apollo-server-express';
import ArticlesDataSource from '../datasources/ArticlesDataSource';

export const ArticlesTypeDefs = gql`
  input ArticleInputArgs {
    title: String!
    body: String!
    userId: Int!
  }

  type Article {
    id: ID
    body: String
    title: String
    user: User
  }

  extend type Query {
    article(id: [Int]): [Article]
    articles: [Article]
  }

  extend type Mutation {
    article(article: ArticleInputArgs!): Article
  }
`;

interface ArticlesContext {
  user: User;
  dataSources: { articles: ArticlesDataSource };
}

interface IArticlesResolversQuery {
  article(
    parent: void,
    args: { id?: number[] },
    context: ArticlesContext,
  ): Promise<Article[]>;
  articles(): Promise<Article[]>;
}

interface IArticlesResolversMutation {
  article(
    parent: void,
    args: { article: { title: string; body: string; userId: number } },
    context: ArticlesContext,
  );
}

interface IArticlesResolvers extends IResolvers {
  Query: IArticlesResolversQuery;
  Mutation: IArticlesResolversMutation;
}

export default class ArticlesResolvers {
  resolvers: IArticlesResolvers;

  constructor() {
    this.resolvers = this.initializeResolvers();
  }

  private initializeResolvers = (): IArticlesResolvers => {
    return {
      Query: {
        article: (_, args, context) => {
          console.log(process.env.NODE_ENV);
          return context.dataSources.articles.get(args.id);
        },
        articles: () => {
          console.log(`get all users`);
        },
      } as IArticlesResolversQuery,
      Mutation: {
        article: (_, args, context) => {
          return context.dataSources.articles.post(args.article);
        },
      },
    };
  };
}
