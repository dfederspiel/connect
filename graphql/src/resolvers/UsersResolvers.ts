import { User } from '@prisma/client';
import { gql, IResolvers } from 'apollo-server-express';
import UserDataSource from '../datasources/UsersDataSource';

export const UsersTypeDefs = gql`
  type User {
    id: ID!
    email: String
    domain: String
  }

  extend type Query {
    user: User
    users: [User]
  }
`;

export default class UsersResolvers {
  resolvers: IResolvers;

  constructor() {
    this.resolvers = this.initializeResolvers();
  }

  initializeResolvers = (): IResolvers => {
    return {
      Query: {
        user: (
          _: any,
          _args: any,
          context: { user: User; dataSources: { userApi: UserDataSource } },
          __: any,
        ) => {
          if (!context.user)
            return {
              error: 'There is no user context, did you forget to pass a bearer token?',
            };
          return context.dataSources.userApi.getById(context.user.id.toString());
        },
        users: (
          _: any,
          _args: any,
          context: { user: User; dataSources: { userApi: UserDataSource } },
          __: any,
        ) => {
          if (!context.user)
            return {
              error: 'There is no user context, did you forget to pass a bearer token?',
            };
          return context.dataSources.userApi.getAll();
        },
      },
    };
  };
}
