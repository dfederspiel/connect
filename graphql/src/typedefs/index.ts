import { gql } from 'apollo-server-express';

export const rootTypeDefs = gql`
  type Query {
    _empty: String
  }

  type Mutation {
    _empty: String
  }

  type Subscription {
    _empty: String
  }
`;
