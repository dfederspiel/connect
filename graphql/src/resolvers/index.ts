import { IResolvers } from '@graphql-tools/utils';
import { IPubSub } from '../lib/types';
import AffirmationsResolvers from './AffirmationsResolvers';
import ArticlesResolvers from './ArticlesResolvers';
import UsersResolvers from './UsersResolvers';

const resolvers = (pubsub: IPubSub): IResolvers[] => [
  new AffirmationsResolvers(pubsub).resolvers,
  new ArticlesResolvers().resolvers,
  new UsersResolvers().resolvers,
];

export default resolvers;
