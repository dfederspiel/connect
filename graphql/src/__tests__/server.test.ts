import GraphQLServer from '../server';
import { PubSub } from 'graphql-subscriptions';
import { gql } from 'apollo-server-express';
import { makeExecutableSchema } from '@graphql-tools/schema';

import { MockContext, Context, createMockContext } from '../../__mocks__/context';
import { AuthContext } from '@lib/auth/AuthContext';
import { rootTypeDefs } from '../typedefs';
import { AffirmationsTypeDefs } from '../resolvers/AffirmationsResolvers';
import { UsersTypeDefs } from '../resolvers/UsersResolvers';
import resolvers from '../resolvers';

jest.spyOn(AuthContext.prototype, 'decode').mockReturnValue(Promise.resolve({}));

const pubsub = new PubSub();
let mockCtx: MockContext;
let ctx: Context;

const schema = makeExecutableSchema({
  typeDefs: [rootTypeDefs, AffirmationsTypeDefs, UsersTypeDefs],
  resolvers: resolvers(pubsub),
});

beforeEach(() => {
  mockCtx = createMockContext();
  ctx = mockCtx as unknown as Context;
});

describe('the graphql server', () => {
  it('exists', () => {
    const server = new GraphQLServer(schema, pubsub, {});
    expect(server).toBeDefined();
  });

  it('can query stuff', async () => {
    mockCtx.prisma.user.findMany.mockResolvedValue([
      {
        id: 1,
        domain: '',
        email: '',
      },
    ]);
    const apollo = new GraphQLServer(schema, pubsub, false, mockCtx.prisma);
    const instance = apollo.server();
    const result = await instance.executeOperation({
      query: gql`
        {
          users {
            id
            email
          }
        }
      `,
    });
    expect(result).toMatchSnapshot();
  });

  it('requires authentication if in production', async () => {
    process.env.NODE_ENV = 'production';
    const apollo = new GraphQLServer(schema, pubsub, false, mockCtx.prisma);
    const instance = apollo.server();
    instance
      .executeOperation({
        query: gql`
          {
            users {
              id
              email
            }
          }
        `,
      })
      .catch((err) => {
        expect(err).toMatchSnapshot();
      });
  });
});
