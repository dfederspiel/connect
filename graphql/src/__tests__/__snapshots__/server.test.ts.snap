// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the graphql server can query stuff 1`] = `
Object {
  "data": Object {
    "users": Array [
      Object {
        "email": "",
        "id": "1",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
